<a href="https://colab.research.google.com/github/Rahul-Chouhan-407/Practice_and_Assignments/blob/main/Decision_Tree_Quiz.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [1]:
import numpy as np
import pandas as pd

In [2]:
df = pd.read_csv('/content/Pokemon.csv')

In [19]:
df.tail(100)

#                      Name    Type 1    Type 2  Total  HP  Attack  \
700  639                 Terrakion      Rock  Fighting    580  91     129   
701  640                  Virizion     Grass  Fighting    580  91      90   
702  641   TornadusIncarnate Forme    Flying      None    580  79     115   
703  641     TornadusTherian Forme    Flying      None    580  79     100   
704  642  ThundurusIncarnate Forme  Electric    Flying    580  79     115   
..   ...                       ...       ...       ...    ...  ..     ...   
795  719                   Diancie      Rock     Fairy    600  50     100   
796  719       DiancieMega Diancie      Rock     Fairy    700  50     160   
797  720       HoopaHoopa Confined   Psychic     Ghost    600  80     110   
798  720        HoopaHoopa Unbound   Psychic      Dark    680  80     160   
799  721                 Volcanion      Fire     Water    600  80     110   

     Defense  Sp. Atk  Sp. Def  Speed  Generation  Legendary  
700       90       72       90    108           5       True  
701       72       90      129    108           5       True  
702       70      125       80    111           5       True  
703       80      110       90    121           5       True  
704       70      125       80    111           5       True  
..       ...      ...      ...    ...         ...        ...  
795      150      100      150     50           6       True  
796      110      160      110    110           6       True  
797       60      150      130     70           6       True  
798       60      170      130     80           6       True  
799      120      130       90     70           6       True  

[100 rows x 13 columns]

In [15]:
df.isnull().sum()

#             0
Name          0
Type 1        0
Type 2        0
Total         0
HP            0
Attack        0
Defense       0
Sp. Atk       0
Sp. Def       0
Speed         0
Generation    0
Legendary     0
dtype: int64

In [13]:
df.shape

(800, 13)

In [14]:
df['Type 2'].fillna("None", inplace=True)

In [16]:
df.describe()

#      Total          HP      Attack     Defense     Sp. Atk  \
count  800.000000  800.00000  800.000000  800.000000  800.000000  800.000000   
mean   362.813750  435.10250   69.258750   79.001250   73.842500   72.820000   
std    208.343798  119.96304   25.534669   32.457366   31.183501   32.722294   
min      1.000000  180.00000    1.000000    5.000000    5.000000   10.000000   
25%    184.750000  330.00000   50.000000   55.000000   50.000000   49.750000   
50%    364.500000  450.00000   65.000000   75.000000   70.000000   65.000000   
75%    539.250000  515.00000   80.000000  100.000000   90.000000   95.000000   
max    721.000000  780.00000  255.000000  190.000000  230.000000  194.000000   

          Sp. Def       Speed  Generation  
count  800.000000  800.000000   800.00000  
mean    71.902500   68.277500     3.32375  
std     27.828916   29.060474     1.66129  
min     20.000000    5.000000     1.00000  
25%     50.000000   45.000000     2.00000  
50%     70.000000   65.000000     3.00000  
75%     90.000000   90.000000     5.00000  
max    230.000000  180.000000     6.00000

In [17]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 800 entries, 0 to 799
Data columns (total 13 columns):
 #   Column      Non-Null Count  Dtype 
---  ------      --------------  ----- 
 0   #           800 non-null    int64 
 1   Name        800 non-null    object
 2   Type 1      800 non-null    object
 3   Type 2      800 non-null    object
 4   Total       800 non-null    int64 
 5   HP          800 non-null    int64 
 6   Attack      800 non-null    int64 
 7   Defense     800 non-null    int64 
 8   Sp. Atk     800 non-null    int64 
 9   Sp. Def     800 non-null    int64 
 10  Speed       800 non-null    int64 
 11  Generation  800 non-null    int64 
 12  Legendary   800 non-null    bool  
dtypes: bool(1), int64(9), object(3)
memory usage: 75.9+ KB


How many pokemon are from the 5th generation?

a. 178
b. 165
c. 150
d. 170

In [20]:
gen_5_pokemon = len(df[df['Generation'] == 5])

print("Number of Pokemon from the 5th generation:", gen_5_pokemon)

Number of Pokemon from the 5th generation: 165


How many pokemon have the highest defense score?

a. 10
b. 7
c. 3
d. 2

In [21]:
# find the maximum defense score
max_defense = df['Defense'].max()

# count the number of Pokemon with the maximum defense score
num_max_defense = len(df[df['Defense'] == max_defense])

print("Number of Pokemon with the highest defense score:", num_max_defense)


Number of Pokemon with the highest defense score: 3


Which columns are not having any kind of relationship with the generation column?

a. Attack
b. Speed
c. Both of the above
d. None of the above

In [22]:

# calculate the correlation coefficients
correlations = df[['Generation', 'HP', 'Attack', 'Defense', 'Sp. Atk', 'Sp. Def', 'Speed']].corrwith(df['Generation'])

print(correlations)


Generation    1.000000
HP            0.058683
Attack        0.051451
Defense       0.042419
Sp. Atk       0.036437
Sp. Def       0.028486
Speed        -0.023121
dtype: float64


In [23]:
df = df.drop(['Name'], axis=1)

In [24]:
df = pd.get_dummies(df, columns=['Type 1', 'Type 2'], drop_first=True)

In [25]:
X = df.drop(['Legendary'], axis=1)
y = df['Legendary']

In [26]:
from sklearn.model_selection import train_test_split

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3, random_state=1)

In [27]:
from sklearn.tree import DecisionTreeClassifier

clf = DecisionTreeClassifier()
clf.fit(X_train, y_train)

DecisionTreeClassifier()

In [28]:
y_pred = clf.predict(X_test)

In [29]:
from sklearn.metrics import accuracy_score, classification_report

print('Accuracy:', accuracy_score(y_test, y_pred))
print('Classification Report:\n', classification_report(y_test, y_pred))

Accuracy: 0.9458333333333333
Classification Report:
               precision    recall  f1-score   support

       False       0.97      0.97      0.97       219
        True       0.70      0.67      0.68        21

    accuracy                           0.95       240
   macro avg       0.83      0.82      0.83       240
weighted avg       0.94      0.95      0.95       240



In [30]:
from sklearn.metrics import confusion_matrix

# Assuming y_test is the actual labels of the test set and y_pred is the predicted labels from the model
tn, fp, fn, tp = confusion_matrix(y_test, y_pred).ravel()

sensitivity = tp / (tp + fn)
print("Sensitivity: ", sensitivity)


Sensitivity:  0.6666666666666666


In [31]:
correctly_classified = tp + tn
print("Correctly classified instances: ", correctly_classified)


Correctly classified instances:  227


In [33]:
from sklearn.metrics import confusion_matrix

y_pred = clf.predict(X_test)
cm = confusion_matrix(y_test, y_pred)
misclassified = cm[0, 1] + cm[1, 0]
print("Number of misclassified data points:", misclassified)


Number of misclassified data points: 13


In [34]:
from sklearn.metrics import confusion_matrix

y_pred = clf.predict(X_test)
cm = confusion_matrix(y_test, y_pred)
misclassified = cm.sum() - cm.diagonal().sum()
print("Misclassified data:", misclassified)


Misclassified data: 13
